python/lsst/ts/dream/csc/dream_csc.py

```diff
@@ -479,6 +525,140 @@ async def send_events(self, status_data: dict[str, Any]) -> None:
+        other = (
+            f"{data_product.start.tai.isot}_{data_product.server}_"
+            f"{data_product.kind}{product_type}_"
+            f"{data_product.seq[0]:06d}_{data_product.seq[-1]:06d}"
+        )
+        key = self.s3bucket.make_key(
+            salname=self.salinfo.name,
+            salindexname=None,
+            generator="dream",
+            date=data_product.start,
+            other=other,
+            suffix=".fits",
+        )
+
+        # Download the object with HTTP
+        dream_url = f"{self.config.url_root}/{data_product.filename}"
+        async with httpx.AsyncClient() as client:
+            async with client.stream("GET", dream_url) as response:
+                if response.status_code != 200:
+                    self.log.error(
+                        f"Data lost [1]: failed to retrieve file: {dream_url} - HTTP {response.status_code}"
+                    )
+                    return
+
+                try:
+                    # First attempt: Save to S3
+                    await self.save_to_s3(response, key)
+                    return  # Success!
+                except Exception:
+                    self.log.exception(
+                        f"Could not upload {key} to S3; trying to save to local disk."
+                    )
+
+            # Second attempt: Fresh request and save locally
+            async with client.stream("GET", dream_url) as response:
+                if response.status_code != 200:
+                    self.log.error(
+                        f"Data lost [2]: failed to retrieve file: {dream_url} - HTTP {response.status_code}"
+                    )
+                    return
+
+                try:
+                    await self.save_to_local_disk(response, key)
+                except Exception:
+                    self.log.exception("Also failed to save file locally. Data lost!")
+
+    async def save_to_s3(self, response: httpx.Response, key: str) -> None:
+        """Upload file to S3 from an HTTP stream.
+
+        Parameters
+        ----------
+        response: httpx.Response
+            HTTP response object for the file to save.
+
+        key: str
+            S3 style filename key to save to.
+        """
+        if not self.s3bucket:
+            raise RuntimeError("S3 bucket not configured")
+
+        try:
+            with io.BytesIO(await response.aread()) as buffer:
+                await self.s3bucket.upload(fileobj=buffer, key=key)
+            url = (
+                f"{self.s3bucket.service_resource.meta.client.meta.endpoint_url}/"
+                f"{self.s3bucket.name}/{key}"
+            )
+            await self.evt_largeFileObjectAvailable.set_write(
+                url=url,
+                generator="dream",
+            )
+            self.log.info(f"Successfully uploaded {key} to S3.")
+        except Exception:
+            self.log.exception(f"Failed to upload {key} to S3.")
```

Comment:
I have the impression you are logging the same exceptions multiple times here. You might want to remove the try/except here and let the exception be lifted and handled at another layer.